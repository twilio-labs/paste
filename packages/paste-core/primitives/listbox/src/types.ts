/**
 * Utils copied from Reakit. Needed because we wrap Composite and CompositeItem and must match their type for `as`.
 * https://github.com/ariakit/ariakit/blob/v1/packages/reakit-utils/src/types.ts#L16
 */
export type As<P = any> = React.ElementType<P>;

export type ListboxPrimitiveVariants = "single" | "multiple";
