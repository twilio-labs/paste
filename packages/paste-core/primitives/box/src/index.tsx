import styled from '@emotion/styled';
import {compose, layout, space, background, border, boxShadow, position, flexbox, system} from 'styled-system';
import css from '@styled-system/css';
import {
  LayoutProps,
  SpaceProps,
  BackgroundProps,
  BorderProps,
  ShadowProps,
  PositionProps,
  FlexboxProps,
  TypographyProps,
} from '@twilio-paste/style-props';
import {
  CursorProperty,
  AppearanceProperty,
  AnimationProperty,
  TransformProperty,
  TransformOriginProperty,
  VisibilityProperty,
  WhiteSpaceProperty,
  UserSelectProperty,
  PointerEventsProperty,
  BoxSizingProperty,
  ResizeProperty,
  TransitionProperty,
  ListStyleTypeProperty,
  ListStylePositionProperty,
  ListStyleImageProperty,
  ObjectFitProperty,
  ObjectPositionProperty,
  BackgroundAttachmentProperty,
  OutlineProperty,
  FloatProperty,
  WillChangeProperty,
} from 'csstype';
import {PseudoPropStyles} from './PseudoPropStyles';
import {BoxPropTypes} from './BoxPropTypes';

export interface BaseBoxProps
  extends React.HTMLAttributes<any>,
    LayoutProps,
    SpaceProps,
    BackgroundProps,
    BorderProps,
    ShadowProps,
    PositionProps,
    FlexboxProps {
  as?: keyof JSX.IntrinsicElements;
  content?: string;
  cursor?: CursorProperty;
  appearance?: AppearanceProperty;
  animation?: AnimationProperty;
  transform?: TransformProperty;
  transformOrigin?: TransformOriginProperty<string>;
  visibility?: VisibilityProperty;
  whiteSpace?: WhiteSpaceProperty;
  userSelect?: UserSelectProperty;
  pointerEvents?: PointerEventsProperty;
  boxSizing?: BoxSizingProperty;
  resize?: ResizeProperty;
  transition?: TransitionProperty;
  listStyleType?: ListStyleTypeProperty;
  listStylePosition?: ListStylePositionProperty;
  listStyleImage?: ListStyleImageProperty;
  objectFit?: ObjectFitProperty;
  objectPosition?: ObjectPositionProperty<string>;
  backgroundAttachment?: BackgroundAttachmentProperty;
  outline?: OutlineProperty<string>;
  float?: FloatProperty;
  willChange?: WillChangeProperty;
  textDecoration?: TypographyProps['textDecoration'];
  /** Typed as any because Box can literally be any HTML element */
  ref?: any | null;
}

interface PseudoStylesProps {
  _after?: BaseBoxProps;
  _before?: BaseBoxProps;
  _focus?: BaseBoxProps;
  _hover?: BaseBoxProps;
  _active?: BaseBoxProps;
  _pressed?: BaseBoxProps;
  _selected?: BaseBoxProps;
  _focusWithin?: BaseBoxProps;
  _invalid?: BaseBoxProps;
  _disabled?: BaseBoxProps;
  _grabbed?: BaseBoxProps;
  _expanded?: BaseBoxProps;
  _checked?: BaseBoxProps;
  _mixed?: BaseBoxProps;
  _odd?: BaseBoxProps;
  _even?: BaseBoxProps;
  _visited?: BaseBoxProps;
  _readOnly?: BaseBoxProps;
  _first?: BaseBoxProps;
  _last?: BaseBoxProps;
  _groupHover?: BaseBoxProps;
  _notFirst?: BaseBoxProps;
  _notLast?: BaseBoxProps;
  _placeholder?: BaseBoxProps;
}

export interface BoxProps extends BaseBoxProps, PseudoStylesProps {}

const extraConfig = system({
  backgroundColor: {
    property: 'backgroundColor',
    scale: 'backgroundColors',
  },
  borderColor: {
    property: 'borderColor',
    scale: 'borderColors',
  },
  animation: true,
  appearance: true,
  transform: true,
  transformOrigin: true,
  visibility: true,
  whiteSpace: true,
  userSelect: true,
  pointerEvents: true,
  boxSizing: true,
  cursor: true,
  resize: true,
  transition: true,
  listStyleType: true,
  listStylePosition: true,
  listStyleImage: true,
  objectFit: true,
  objectPosition: true,
  backgroundAttachment: {
    property: 'backgroundAttachment',
  },
  outline: true,
  float: true,
  willChange: true,
  textDecoration: true,
});

const getPseudoStyles = (props: BoxProps): {} => {
  const pseudoProps = Object.keys(props).filter(propName => propName.startsWith('_'));

  if (pseudoProps.length === 0) {
    return {};
  }

  const pseudoStyles = {};
  pseudoProps.forEach(pseudoProp => {
    if (PseudoPropStyles[pseudoProp] != null) {
      pseudoStyles[PseudoPropStyles[pseudoProp]] = props[pseudoProp];
    }
  });

  return css(pseudoStyles);
};

/* eslint-disable emotion/syntax-preference */
export const Box = styled.div(
  {
    boxSizing: 'border-box',
  },
  compose(
    space,
    layout,
    flexbox,
    background,
    border,
    boxShadow,
    position,
    extraConfig
  ),
  getPseudoStyles
) as React.FC<BoxProps>;
/* eslint-enable */

Box.displayName = 'Box';

if (process.env.NODE_ENV === 'development') {
  Box.propTypes = BoxPropTypes;
}

export * from './SafelySpreadProps';
