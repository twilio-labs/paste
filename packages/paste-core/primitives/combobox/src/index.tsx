import Downshift from '@twilio-paste/dropdown-library';

export {Downshift as ComboboxPrimitive};

export {useCombobox as useComboboxPrimitive} from '@twilio-paste/dropdown-library';

export type {
  UseComboboxInterface as UseComboboxPrimitiveInterface,
  UseComboboxProps as UseComboboxPrimitiveProps,
  UseComboboxState as UseComboboxPrimitiveState,
  UseComboboxReturnValue as UseComboboxPrimitiveReturnValue,
  UseComboboxStateChange as UseComboboxPrimitiveStateChange,
} from '@twilio-paste/dropdown-library';
