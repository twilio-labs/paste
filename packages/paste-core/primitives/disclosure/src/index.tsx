// https://reakit.io/docs/disclosure/
export {
  useDisclosureState as useDisclosurePrimitiveState,
  Disclosure as DisclosurePrimitive,
  DisclosureContent as DisclosurePrimitiveContent,
} from '@twilio-paste/reakit-library';
