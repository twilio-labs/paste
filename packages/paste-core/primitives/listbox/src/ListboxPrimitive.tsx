import type { CompositeProps } from "@twilio-paste/reakit-library";
import { Composite } from "@twilio-paste/reakit-library";
import * as React from "react";

import type { As, ListboxPrimitiveVariants } from "./types";

export interface ListboxPrimitiveProps
  extends Omit<
    CompositeProps,
    "unstable_virtual" | "unstable_moves" | "unstable_system" | "wrapElement" | "wrap" | "as"
  > {
  variant?: ListboxPrimitiveVariants;
  as?: As;
}

export const ListboxPrimitive = React.forwardRef<HTMLDivElement, ListboxPrimitiveProps>(
  ({ variant = "single", orientation, ...props }, ref) => (
    <Composite
      as="div"
      {...props}
      ref={ref}
      role="listbox"
      aria-multiselectable={variant === "multiple"}
      aria-orientation={orientation}
    />
  ),
);

ListboxPrimitive.displayName = "ListboxPrimitive";
