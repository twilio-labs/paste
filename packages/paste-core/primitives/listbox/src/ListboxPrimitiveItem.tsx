import * as React from 'react';
import type {CompositeItemProps} from '@twilio-paste/reakit-library';
import {CompositeItem} from '@twilio-paste/reakit-library';

import type {As} from './types';

export interface ListboxPrimitiveItemProps
  extends Omit<CompositeItemProps, 'unstable_virtual' | 'unstable_moves' | 'unstable_system' | 'wrapElement' | 'wrap'> {
  selected?: boolean;
  /** Event handler to respond to selection events */
  onSelect?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  as?: As;
}

export const ListboxPrimitiveItem = React.forwardRef<HTMLButtonElement, ListboxPrimitiveItemProps>(
  ({selected = false, onSelect, ...props}, ref) => (
    <CompositeItem
      as="button"
      {...props}
      ref={ref}
      role="option"
      aria-selected={selected}
      type="button"
      onClick={onSelect}
    />
  )
);

ListboxPrimitiveItem.displayName = 'ListboxPrimitiveItem';
