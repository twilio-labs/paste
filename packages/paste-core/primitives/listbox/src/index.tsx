export * from "./ListboxPrimitive";
export * from "./ListboxPrimitiveItem";
export * from "./useListboxPrimitiveState";

export { CompositeGroup as ListboxPrimitiveGroup } from "@twilio-paste/reakit-library";
export type { CompositeGroupProps as ListboxPrimitiveGroupProps } from "@twilio-paste/reakit-library";
