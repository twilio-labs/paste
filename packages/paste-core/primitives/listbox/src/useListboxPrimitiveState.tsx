import { useCompositeState } from "@twilio-paste/reakit-library";
import type {
  CompositeInitialState,
  CompositeStateReturn as ListboxPrimitiveStateReturn,
} from "@twilio-paste/reakit-library";

export type { ListboxPrimitiveStateReturn };
export type ListboxPrimitiveInitialState = Omit<CompositeInitialState, "loop">;

export const useListboxPrimitiveState = (config: ListboxPrimitiveInitialState = {}): ListboxPrimitiveStateReturn => {
  return {
    ...useCompositeState({
      orientation: "vertical", // vertical by default
      ...config,
      loop: "horizontal", // loops when orientation is horizontal
    }),
  };
};
