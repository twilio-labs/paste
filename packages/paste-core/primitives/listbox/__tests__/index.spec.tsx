import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as React from "react";

import { ListboxPrimitive, ListboxPrimitiveItem, useListboxPrimitiveState } from "../src";

const FIRST_ITEM = "first-listbox-item";
const SECOND_ITEM = "second-listbox-item";

const ListboxMock: React.FC = () => {
  const [selected, setSelected] = React.useState<number>();
  const listbox = useListboxPrimitiveState({ baseId: "listbox-example" });
  return (
    <ListboxPrimitive {...listbox} aria-label="Code">
      <ListboxPrimitiveItem
        {...listbox}
        data-testid={FIRST_ITEM}
        selected={selected === 1}
        onSelect={() => {
          setSelected(1);
        }}
      >
        Item 1
      </ListboxPrimitiveItem>
      <ListboxPrimitiveItem
        {...listbox}
        data-testid={SECOND_ITEM}
        selected={selected === 2}
        onSelect={() => {
          setSelected(2);
        }}
      />
    </ListboxPrimitive>
  );
};

const ListboxHorizontalMock: React.FC = () => {
  const listbox = useListboxPrimitiveState({ baseId: "listbox-horizontal-example", orientation: "horizontal" });
  return (
    <ListboxPrimitive {...listbox} aria-label="Code">
      <ListboxPrimitiveItem {...listbox} data-testid={FIRST_ITEM}>
        Item 1
      </ListboxPrimitiveItem>
      <ListboxPrimitiveItem {...listbox} data-testid={SECOND_ITEM} />
    </ListboxPrimitive>
  );
};

const ListboxMultipleMock: React.FC = () => {
  const [selectedSet, updateSelectedSet] = React.useState<Set<number>>(new Set());
  const listbox = useListboxPrimitiveState({ baseId: "listbox-multi-example" });
  const updateSelection = (item: number): void => {
    const newSelectedSet = new Set(selectedSet);
    if (newSelectedSet.has(item)) {
      newSelectedSet.delete(item);
    } else {
      newSelectedSet.add(item);
    }
    updateSelectedSet(newSelectedSet);
  };
  return (
    <ListboxPrimitive {...listbox} aria-label="Code" variant="multiple">
      <ListboxPrimitiveItem {...listbox} selected={selectedSet.has(1)} onSelect={() => updateSelection(1)}>
        Item 1
      </ListboxPrimitiveItem>
      <ListboxPrimitiveItem {...listbox} selected={selectedSet.has(2)} onSelect={() => updateSelection(2)} />
    </ListboxPrimitive>
  );
};

describe("Listbox Primitive", () => {
  describe("Render", () => {
    it("should render", () => {
      render(<ListboxMock />);
      expect(screen.getByRole("listbox")).toBeVisible();
    });

    it("should render a listbox", () => {
      render(<ListboxMock />);
      const renderedListbox = screen.getByLabelText("Code");
      expect(renderedListbox.getAttribute("role")).toEqual("listbox");
      expect(renderedListbox.getAttribute("aria-orientation")).toEqual("vertical");
    });

    it("should render a horizontal listbox", () => {
      render(<ListboxHorizontalMock />);
      const renderedListbox = screen.getByLabelText("Code");
      expect(renderedListbox.getAttribute("aria-orientation")).toEqual("horizontal");
    });

    it("should render a listbox item", () => {
      render(<ListboxMock />);
      const renderedListboxItem = screen.getByTestId(FIRST_ITEM);
      expect(renderedListboxItem.getAttribute("role")).toEqual("option");
    });

    it("should render a multiselect listbox", () => {
      render(<ListboxMultipleMock />);
      const renderedListbox = screen.getByLabelText("Code");
      expect(renderedListbox.getAttribute("aria-multiselectable")).toEqual("true");
    });
  });

  describe("Interaction", () => {
    it("should move focus to the second item when down arrow is pressed", () => {
      render(<ListboxMock />);
      userEvent.tab();
      userEvent.keyboard("{arrowdown}");
      expect(screen.getByTestId(SECOND_ITEM)).toHaveFocus();
    });

    it("should move focus to the first item when down and up is pressed", () => {
      render(<ListboxMock />);
      userEvent.tab();
      userEvent.keyboard("{arrowdown}");
      userEvent.keyboard("{arrowup}");
      expect(screen.getByTestId(FIRST_ITEM)).toHaveFocus();
    });

    it("should move focus to the second item when right arrow is pressed for horizontal", () => {
      render(<ListboxHorizontalMock />);
      userEvent.tab();
      userEvent.keyboard("{arrowright}");
      expect(screen.getByTestId(SECOND_ITEM)).toHaveFocus();
    });

    it("should move focus to the first item when right and left is pressed for horizontal", () => {
      render(<ListboxHorizontalMock />);
      userEvent.tab();
      userEvent.keyboard("{arrowright}");
      userEvent.keyboard("{arrowleft}");
      expect(screen.getByTestId(FIRST_ITEM)).toHaveFocus();
    });

    it("should loop focus to the first item when right and right is pressed for horizontal", () => {
      render(<ListboxHorizontalMock />);
      userEvent.tab();
      userEvent.keyboard("{arrowright}");
      userEvent.keyboard("{arrowright}");
      expect(screen.getByTestId(FIRST_ITEM)).toHaveFocus();
    });

    it("should select the first item when enter is pressed", () => {
      render(<ListboxMock />);
      userEvent.tab();
      userEvent.keyboard("{enter}");

      const firstListboxItem = screen.getByTestId(FIRST_ITEM);
      expect(firstListboxItem.getAttribute("aria-selected")).toEqual("true");
      const secondListboxItem = screen.getByTestId(SECOND_ITEM);
      expect(secondListboxItem.getAttribute("aria-selected")).toEqual("false");
    });

    it("should select both items for a multiselect listbox", () => {
      render(<ListboxMultipleMock />);
      userEvent.tab();
      userEvent.keyboard("{enter}");
      userEvent.keyboard("{arrowdown}");
      userEvent.keyboard("{enter}");

      const options = screen.getAllByRole("option");
      expect(options[0].getAttribute("aria-selected")).toEqual("true");
      expect(options[1].getAttribute("aria-selected")).toEqual("true");
    });
  });
});
