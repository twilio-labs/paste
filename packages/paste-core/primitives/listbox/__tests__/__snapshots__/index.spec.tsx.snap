// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listbox Primitive Render should render 1`] = `
<DocumentFragment>
  <div
    aria-label="Code"
    aria-multiselectable="false"
    aria-orientation="vertical"
    id="listbox-example"
    role="listbox"
  >
    <button
      aria-selected="false"
      data-testid="first-listbox-item"
      id="listbox-example-1"
      role="option"
      tabindex="0"
    >
      Item 1
    </button>
    <button
      aria-selected="false"
      data-testid="second-listbox-item"
      id="listbox-example-2"
      role="option"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;
